uniffi/src/lib.rs

```diff
@@ -629,6 +629,67 @@ unsafe impl<T: Sync + Send> FfiConverter for std::sync::Arc<T> {
     }
 }
 
+/// Support for passing reference-counted trait objects via the FFI.
+///
+/// This is different than the last one, since trait objects are stored as 2 pointers (AKA a "fat
+/// pointer): one for `T` and another for the vtable.  Since the foreign side is expecting
+/// a single pointer, something needs to get wrapped in a Box. And all our `T`s are `Arc<>`s.
+/// So, of `Box<Arc<T>` or `Arc<Box<T>`, an inner `Arc<T>` is more consistent with other impls.
+unsafe impl<T: Sync + Send + ?Sized> FfiConverter for Box<std::sync::Arc<T>> {
+    type RustType = std::sync::Arc<T>;
+    // Don't use a pointer to <T> as that requires a `pub <T>`
+    type FfiType = *const std::os::raw::c_void;
+
+    /// When lowering, we have an owned `Arc<T>` and we transfer that ownership
+    /// to the foreign-language code, "leaking" it out of Rust's ownership system
+    /// as a raw pointer. This works safely because we have unique ownership of `self`.
+    /// The foreign-language code is responsible for freeing this by calling the
+    /// `ffi_object_free` FFI function provided by the corresponding UniFFI type.
+    ///
+    /// Safety: when freeing the resulting pointer, the foreign-language code must
+    /// call the destructor function specific to the type `T`. Calling the destructor
+    /// function for other types may lead to undefined behaviour.
+    fn lower(obj: Self::RustType) -> Self::FfiType {
+        Box::into_raw(Box::new(obj)) as *const std::os::raw::c_void
+    }
+
+    /// When lifting, we receive a "borrow" of the `Arc<T>` that is owned by
+    /// the foreign-language code, and make a clone of it for our own use.
+    ///
+    /// Safety: the provided value must be a pointer previously obtained by calling
+    /// the `lower()` or `write()` method of this impl.
+    fn try_lift(v: Self::FfiType) -> Result<Self::RustType> {
+        // Use Box::leak to avoid dropping the `Arc<T>` that is owned by the foreign-language code.
+        let foreign_arc = Box::leak(unsafe { Box::from_raw(v as *mut std::sync::Arc<T>) });
```

Comment:
And I guess this is where the leak comes from?  This really makes me think `Arc<Box<T>>` is the way to go and maybe it's more precise to call this `Arc<Box<dyn Trait>>`.  I think I can even wrap my head around it because it's closer the the way we both intuitively thought that things worked.  `Box<dyn Trait>` is a fat pointer, where one points to the underlying T and the other points to the vtable.  Then `Arc<Box<dyn trait>>` let's you share that between the foreign and Rust code, with the nice side benefit it's just a regular, skinny, pointer.

In fact, could you then just use the existing `Arc<T>` impl?

Reply:
I'm still a bit confused by this TBH - I think I need to better understand the leak.

My only real problem with `Arc<Box<dyn T>>` is the ergonomics, and I'm not even sure how much of a problem I think it is :) But having an argument be `foo: Arc<Box<impl Foo>>` seems like it's going to wind up with lots of `&**foo` and other gymnastics which makes the code look messy and just surfaces the limitations of UniFFI rather than looking "natural".

So I'm still trying to think about ways we can avoid this and use the slightly more natural `Arc<impl Foo>` in the signatures of functions being exposed. Although we discussed how ugly and hacky it would be, and I expressed hostility to the idea, I am starting to wonder how bad it would be to try and pull some of the tricks mentioned in [this discussion about `std::raw::TraitObject`](https://github.com/rust-lang/rust/issues/27751)? Or whether we can make the current model not leak (which seems like it *should* be possible, although might end up being very heavy on many small, short-lived allocations). Or something :) WDYT?

> I think whatever way we choose to implement this, we need good documentation about how everything is layed out.

Agreed - mermaid FTW I guess :)

Reply:
> I think I need to better understand the leak.

I can't actually reproduce a leak in that lower/lift implementation - there is a leak in my branch, but it doesn't look like it's there - so let me get more info before we start exploring the more radical options.

Reply:
> My only real problem with `Arc<Box<dyn T>>` is the ergonomics, and I'm not even sure how much of a problem I think it is :) But having an argument be `foo: Arc<Box<impl Foo>>` seems like it's going to wind up with lots of `&**foo` and other gymnastics which makes the code look messy and just surfaces the limitations of UniFFI rather than looking "natural".

That's a good point.  I wonder if we could write our own wrapper type that uses `Arc<Box<dyn T>>` internally and has `Deref` impl that makes it easy to use.  Or maybe the type should wrap `Box<dyn T>` in a way that makes it easy to use when wrapped in an Arc.